src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -152,65 +155,103 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
             if (!existingPackageReferences.Any())
             {
-                // Add packageReference only if it does not exist.
+                // Getting all the item groups in a given project
+                var itemGroups = GetItemGroups(project);
+
+                // Add packageReference to the project file only if it does not exist.
                 var itemGroup = GetItemGroup(itemGroups, PACKAGE_REFERENCE_TYPE_TAG) ?? CreateItemGroup(project, framework);
-                AddPackageReferenceIntoItemGroup(itemGroup, libraryDependency);
+
+                // If onboarded to CPM props file must be updated. 
+                if (libraryDependency.VersionCentrallyManaged)
+                {
+                    // Modifying the props file if project is onboarded to CPM.
+                    UpdatePropsFileWithNewPackageReference(project, libraryDependency);
+                    //Modify the project file.
+                    AddPackageReferenceIntoItemGroupCPM(itemGroup, libraryDependency);
+                }
+                else
+                {
+                    //Modify the project file.
+                    AddPackageReferenceIntoItemGroup(itemGroup, libraryDependency);
+                }
             }
             else
             {
-                // If the package already has a reference then try to update the reference.
-                UpdatePackageReferenceItems(existingPackageReferences, libraryDependency);
+                if (libraryDependency.VersionCentrallyManaged)
+                {
+                    UpdatePropsFileWithPackageVersion(project, libraryDependency);
+                }
+                else
+                {
+                    // If the package already has a reference then try to update the reference.
+                    UpdatePackageReferenceItems(existingPackageReferences, libraryDependency);
+                }
             }
+
             project.Save();
         }
 
-        private static IEnumerable<ProjectItemGroupElement> GetItemGroups(Project project)
+        private void UpdatePropsFileWithNewPackageReference(Project project, LibraryDependency libraryDependency)
         {
-            return project
-                .Items
-                .Where(i => !i.IsImported)
-                .Select(item => item.Xml.Parent as ProjectItemGroupElement)
-                .Distinct();
+            // If onboarded to CPM get the directoryBuildPropsRootElement.
+            ProjectRootElement directoryBuildPropsRootElement = GetDirectoryBuildPropsRootElement(project);
+
+            // Get the ItemGroup to add a PackageVersion to or create a new one.
+            var propsItemGroup = GetItemGroup(directoryBuildPropsRootElement.ItemGroups, PACKAGE_VERSION_TYPE_TAG) ?? directoryBuildPropsRootElement.AddItemGroup();
+            AddPackageReferenceIntoPropsItemGroup(propsItemGroup, libraryDependency);
+
+            // Save the updated props file.
+            directoryBuildPropsRootElement.Save();
```

Comment:
Since we are no longer updating the package version in the props file and rather adding a VersionOverride attribute to the project file instead, I think this no longer needs to be addressed. But please let me know if there are similar concerns with updating the VersionOverride attribute (I don't think there are because we are updating the project file not the props file, but I could be wrong).

Reply:
Consider using [`ProjectRootElement.PreserveFormatting`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.construction.projectrootelement.preserveformatting?view=msbuild-17-netcore) to preserve formatting. I didn't try this API :)